### PR TITLE
Overwrite old node labels for first node update to the endpoint manager (v1.14)

### LINKED ISSUE #29649: Node labels updates rejected indefinitely by host endpoint if node watcher initializes before the host endpoint is created
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

We ran into a race condition where the host endpoint's labels are outdated indefinitely. The issue is fixed if I restart the cilium agent.

The sequence of steps that resulted in the race condition are:

1. The endpoint manager's node label observer is initialized before the host endpoint is created. So, the first node label update is missed as we return early if the host endpoint is not found.

https://github.com/cilium/cilium/blob/d1128cf948979de7433385b2101a98404d99c64f/pkg/endpointmanager/host.go#L43-L48

2. All the subsequent node label updates are rejected by endpoint's `ModifyIdentityLabels` function

https://github.com/cilium/cilium/blob/d1128cf948979de7433385b2101a98404d99c64f/pkg/labels/oplabels.go#L157-L169

As the first node label update was missed, some labels on the oldNode object are not found on the endpoint's existing labels so the update request is rejected.

### Cilium Version

All versions

### Kernel Version

Linux bootstrapper 5.4.0-100-generic

### Kubernetes Version

1.27

### Sysdump

_No response_

### Relevant log output

_No response_

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by skmatti**
This fixes a race condition during agent startup where some node label updates were missed. If the k8s node watcher initializes before the host endpoint is created, the agent might miss some node label updates.

Fixes #29649

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

Once this PR is merged, a GitHub action will update the labels of these PRs:
```upstream-prs
 33551
```

### CODE REVIEW

File: pkg/endpoint/endpoint.go
Comment (pchaigno) [inline pkg/endpoint/endpoint.go]: *(Not on a specific line of code on purpose.)*

I'd include more explanations in the commit description. Right now, your issue has a better explanation than the commit and I believe it would currently be hard for someone without context to understand your pull request. Reviewers are likely to struggle because of that.

One way to check this is to ask a colleague without context to review. If they fail to do so without reading the issue or asking you questions, then the commits probably don't have enough information.

Comment (skmatti) [reply pkg/endpoint/endpoint.go]: Thanks, I just updated the commit description. WDYT?

**Review by squeed — CHANGES_REQUESTED** [review]
This doesn't seem like the right way to fix this.

Rather, the issue is that we're passing the wrong information to `updateHostEndpointLabels()`.

My suggestion:

```go
// in ./pkg/endpointmanager/host.go
func (mgr *endpointManager) startNodeLabelsObserver(old map[string]string) {
	mgr.localNodeStore.Observe(context.Background(), func(ln node.LocalNode) {
		if maps.Equal(old, ln.Labels) {
			return
		}

		if err := mgr.updateHostEndpointLabels(old, ln.Labels); err != nil {
			log.Error(err)
		} else {
			old = ln.Labels
		}
	}, func(error) { /* Executed only when we are shutting down */ })
}

func (mgr *endpointManager) updateHostEndpointLabels(oldNodeLabels, newNodeLabels map[string]string) error {
	nodeEP := mgr.GetHostEndpoint()
	if nodeEP == nil {
		return fmt.Errorf("host endpoint not found")
	}

	err := nodeEP.UpdateLabelsFrom(oldNodeLabels, newNodeLabels, labels.LabelSourceK8s)
	if err != nil {
		// An error can only occur if either the endpoint is terminating, or the
		// old labels are not found. Both are impossible, hence there's no point
		// in retrying.
		return fmt.Errorf("unable to update host endpoint labels: %w", err)
	}
	return nil
}
```

Otherwise we will still miss edges, since we update "old" without actually successfully applying. Make sense?

**Comment by learnitall** [comment]
>I think this issue is fixed in 1.15+ and exists only in 1.14 and below:

Would you be able to do more testing to verify this? If this fix is only required for 1.14 and below, then we'd have to change the base of the PR and the target branch the PR merges into. Additionally, it'll help inform discussion on how the fix works.

**Comment by joestringer** [comment]
Per @learnitall's comments above, it seems like the approach needs more discussion. Feel free to rework this PR to base against v1.14 and make a concrete proposal, or create a new PR against that branch, or reach out in #sig-policy on Slack to discuss more free-form about how to approach this.

In the mean time, I'll mark this PR as draft while we figure out a path forward.

**Comment by skmatti** [comment]
I rebased the change to 1.14 only. Also, I made a new fix to address the issue.

I added a unit test as well

**Comment by joestringer** [comment]
Although the CODEOWNERS file requests @cilium/tophat approval, given this is a net-new PR for v1.14 it should really be reviewed by @cilium/endpoint and/or @cilium/sig-policy . I've requested @squeed's fresh review given he also previously looked at it.

**Comment by joestringer** [comment]
/test-backport-1.14

**Comment by joestringer** [comment]
It would be good to evaluate whether this also suits a v1.13 backport once approved.

**Review by pchaigno — APPROVED** [review]

File: pkg/endpointmanager/host.go
Comment (squeed) [inline pkg/endpointmanager/host.go]: nit: pass `nil` here, handle nil below. Otherwise, this looks a lot better.

**Comment by joestringer** [comment]
Almost all tests seem to be failing because this PR is causing the following log message:

> time="2024-06-24T22:03:39Z" level=error msg="Host endpoint not found" subsys=endpoint-manager (1 occurrences)

File: pkg/endpointmanager/host.go
Comment (skmatti) [reply pkg/endpointmanager/host.go]: Done

**Comment by skmatti** [comment]
> Almost all tests seem to be failing because this PR is causing the following log message:
> 
> > time="2024-06-24T22:03:39Z" level=error msg="Host endpoint not found" subsys=endpoint-manager (1 occurrences)

The missing host endpoint error is now being logged on OnNodeAdd as well. One option is to change this log to warning but I think it feels like something that needs be handled in e2e test logic.

**Comment by joestringer** [comment]
@skmatti Those warnings are usually put in place to highlight that there's a problem with the state that the agent is trying to handle, and the solution is to rework the code so that this codepath doesn't get hit in this unexpected condition. Unless I see a clear description otherwise, this looks like a code bug and not a CI issue. Similarly, at a glance I am skeptical that lowering the severity of this warning will address the underlying problem. What is the user supposed to do if they hit this condition? It seems like Cilium is just operating incorrectly.

**Comment by skmatti** [comment]
> @skmatti Those warnings are usually put in place to highlight that there's a problem with the state that the agent is trying to handle, and the solution is to rework the code so that this codepath doesn't get hit in this unexpected condition. Unless I see a clear description otherwise, this looks like a code bug and not a CI issue. Similarly, at a glance I am skeptical that lowering the severity of this warning will address the underlying problem. What is the user supposed to do if they hit this condition? It seems like Cilium is just operating incorrectly.

Host not found errors are expected when the watcher is initialized before the host endpoint is created. The expectation is that the host endpoint will be created with correct labels and new label updates are consumed by this workflow. Let me try to rework to avoid this error

**Comment by skmatti** [comment]
I noticed that a fix is also needed at main for this so I created https://github.com/cilium/cilium/pull/33511

**Review by dylandreimerink — APPROVED** [review]

**Comment by joestringer** [comment]
Given https://github.com/cilium/cilium/pull/33511 is now merged, I think that this PR should pass CI once it is rebased?

**Comment by joestringer** [comment]
I missed that this was the v1.14 version, thanks for the update. I re-requested @squeed's review.

**Comment by joestringer** [comment]
I still see the `level=error msg="Host endpoint not found"` error in CI. Does this need tweaking further?

**Comment by skmatti** [comment]
> I still see the `level=error msg="Host endpoint not found"` error in CI. Does this need tweaking further?

These errors are always happening but the `OnAddNode` method did not log it earlier. 
https://github.com/cilium/cilium/blob/8acae59519adb5e17b7138e39db3ad8e55519ab2/pkg/endpointmanager/host.go#L42-L46

I decoupled the `OnAddNode` and `OnUpdateNode` methods so the error log is gone now but the error is still happening. However, this should not be causing any failures.

@squeed can you take another pass at this?.

**Comment by joestringer** [comment]
What do you mean by "the error is still happening. However, this should not be causing any failures."?

As a general rule, we try to avoid Cilium posting inconsequential errors into the logs, because these can confuse users into thinking something is wrong, or even in an outage scenario, confuse users into thinking that this issue is related to their problem when it may not be. It would be nice to avoid the error overall if we can.

**Comment by skmatti** [comment]
> What do you mean by "the error is still happening. However, this should not be causing any failures."?

There two asynchronous events happening on the agent start up. One is to create the host endpoint. Second one is k8s node object watch logic. There is no guarantee that one comes after the other. If k8s node object watcher is initialized before the host endpoint is created, the host endpoint labels cannot be updated and we bail out early. The problem with current logic is that we do not log an error on `OnAddNode` but we do on `OnUpdateNode` if the host endpoint is not found. 

I think `host endpoint not found` should not logged at error log-level as it is not indicative of a real failure. Even if node labels update is missed, the host endpoint will be created with correct labels and future label updates will be consumed through the node watcher logic.

**Comment by pchaigno** [comment]
@skmatti Do I understand correctly that this bug was introduced in commit d06563e46ca, when the debug log was made into an error log and the next line (`node.SetLabels(newNodeLabels)`) was dropped?

cc @ldelossa

**Comment by joestringer** [comment]
> I think host endpoint not found should not be logged at error log-level as it is not indicative of a real failure. Even if node labels update is missed, the host endpoint will be created with correct labels and future label updates will be consumed through the node watcher logic.

As long as we can ensure the latter, I agree it doesn't make sense to log this an error level (or really anything above debug level since it's an expected condition).

**Comment by joestringer** [comment]
/test-backport-1.14

**Comment by skmatti** [comment]
> @skmatti Do I understand correctly that this bug was introduced in commit [d06563e](https://github.com/cilium/cilium/commit/d06563e46ca58f8974398ace4d144bc9d91d436e), when the debug log was made into an error log and the next line (`node.SetLabels(newNodeLabels)`) was dropped?
> 
> cc @ldelossa

I agree, removing the `node.SetLabels(newNodeLabels)` prevents the host endpoint to be created with latest labels. This was added later but the placement of it is incorrect

**Review by squeed — APPROVED** [review]

File: pkg/endpointmanager/host.go
Comment (squeed) [inline pkg/endpointmanager/host.go:70]: Assuming this error message doesn't show up in regular lifecycle churn, I'm OK with this being an error. That said, the lifecycles of nodes and endpoints are theoretically independent -- I'm not sure what the invariants are here.

Comment (skmatti) [reply pkg/endpointmanager/host.go:70]: Yeah, this does not show up in regular lifecycle churn. This error message indicate a real failure if the node labels are not updated and host firewall could not applied correctly. The downstream does not log this error so it was difficult to debug this issue in our env.

**Comment by pchaigno** [comment]
/test-backport-1.14

**Review by joestringer — APPROVED** [review]

**Pull request merged**
